stan/math/prim/meta/scalar_type.hpp

```diff
@@ -0,0 +1,64 @@
+#ifndef STAN_MATH_PRIM_META_SCALAR_TYPE_HPP
+#define STAN_MATH_PRIM_META_SCALAR_TYPE_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/meta/is_eigen.hpp>
+#include <stan/math/prim/meta/value_type.hpp>
+#include <stan/math/prim/meta/is_vector.hpp>
+#include <stan/math/prim/meta/scalar_type.hpp>
+#include <type_traits>
+#include <vector>
+
+namespace stan {
+
+/** \ingroup type_trait
+ * End of recursion to determine base scalar type of a type
+ * The underlying base scalar type. If T is not a container then this
+ * has a static member nameed type with the type T.
+ * @tparam T the type.
+ */
+template <typename T, typename = void>
+struct scalar_type_base {
+  using type = T;
+};
+
+/** \ingroup type_trait
+ * Metaprogram structure to determine the base scalar type
+ * of a template argument.
+ *
+ * <p>This base class should be specialized for structured types.</p>
+ *
+ * @tparam T Type of object.
+ */
+template <typename T, typename = void>
+struct scalar_type {
+  using type = typename scalar_type_base<std::remove_cv_t<T>>::type;
+};
+
+template <typename T>
+using scalar_type_t = typename scalar_type<T>::type;
+
+template <typename T>
+using scalar_type_decay_t = typename scalar_type<std::decay_t<T>>::type;
+
+/** \ingroup type_trait
+ * Specialization of scalar_type for vector to recursivly return the inner
+ * scalar type.
+ */
+template <typename T>
+struct scalar_type<T, std::enable_if_t<is_std_vector<T>::value>> {
```

Comment:
The arr/meta/scalar_type previously used require generics here (require_std_vector_t). But scalar_type is used in require_generics.hpp and that made scalar_type a dependency of require_generics and vice_versa. I used the type traits directly here to fix this. Not sure there is a better way. If there is I am all for it.

cc @SteveBronder

Reply:
@rok-cesnovar --- I'm reviewing this now and can't understand this comment.  Is there something that needs to change before this gets merged?  My understanding is that this PR is just rearranging where code is at in terms of files, not changing behavior.

Reply:
I was just pointing this out as these 2 changes are the only 2 where I didnt simply move around code from other files. All the additions are just moves from their current respective /arr/ or /mat/. 

Here I had to actually make changes to make the flattening work.